add clipboard support to input capture portal

sithlord48 opened the pull request:
hook the clipboard to input capture portal 

 - [X] Finalized XML bits
 - [X] Ported clipboard parts from remotedesktop portal
 - [ ] Proper `handle_create_draft_session`
 - [ ] Proper `handle_init_session` 
 - [ ] have old `handle_capture_session`  call  `handle_create_draft_session` then `handle_init_session`

Taken over by #1803

jadahl commented:
This was discussed a bit on Matrix, and we need to make a decision how to integrate with the clipboard portal here.

The input capture portal has the `CreateSession()` method as the main entry which queries the user etc (it is the only method with `parent_window`). However the existing way to integrate clipboard access with another portal (the remote desktop portal) is to `RequestClipboard()` on an exsiting session before it's "started", where starting is assumed to be the action that takes a `parent_window` and potentially queries the user.

It is important, however, that `RequestClipboard()` happens before the user is queried, so nagging only potentially happens once, as is the case with the clipboard portal. Thus, I see two options of how to deal with this:

1. Split up `org.freedesktop.portal.InputCapture.CreateSession()` into two parts: `CreateSession2()` (with a better name) and `InitSession()`. The `CreateSession2()` method would practically do nothing more than creating a dormant session that doesn't do anything, and `Init()` would carry the `parent_window` and carry all the parameters that `CreateSession()` has. `CreateSession()` would be a short hand for `CreateSession2(); InitSession();`. This way, clipboard integration can be done in a consistent manner, i.e. `CreateSession2(); RequestClipboard(); InitSession();`.
2. Make clipboard requesting part of `CreateSession()`, e.g. via a `request_clipboard` option in the vararg. This has the benefit of not having to split up `CreateSession()`, but also makes clipboard integration inconsistent and messy, e.g. tricker to carry the (so far unused `options` vararg of `RequestClipboard` and any semantics.

Cc: @whot, do you have any preference? Personally I'm somewhat more fond of option 1, assuming one can come up with a better name.

whot commented:
what may work: Instead of splitting you could have a `draft` boolean option in `CreateSession` that tells it to create a draft session that is later initialized with `Init()`? This way we don't need to split them up, I think. But otherwise I agree, option 1 seems like the more correct one.

sithlord48 commented:
is there somewhere i can see an exmple of a draft option ? feel free to ping me on matrix or respond here.

jadahl commented:
> what may work: Instead of splitting you could have a draft boolean option in CreateSession that tells it to create a draft session that is later initialized with Init()? This way we don't need to split them up, I think. But otherwise I agree, option 1 seems like the more correct one.

A problem I see with that is that the CreateSession is a "request" that takes a parent_window, and it's a bit unorthodox to make an assumption that such a parent window is valid after the request was finished. As in, the potentially temporary resource "parent_window" would not be used until `Init()`.

whot commented:
An empty string for a parent window is valid, right? So the combination of empty parent window and draft set is sufficient (empty parent window can be a requiremen if draft is set?)

> is there somewhere i can see an exmple of a draft option ? feel free to ping me on matrix or respond here.

To copy/paste from how the documentation would look like:

```
 Supported keys in the options vardict include:

   draft (b)

   If true the created session is dormant until a subsequent call to Init(). This allows a caller 
   to invoke other functions on the session before prompting the user, e.g. RequestClipboard().
```
IOW it's just a boolean shoved into the `options` `a{sv}`.

sithlord48 commented:
@whot  is that update correct in the xml  ?

whot commented on data/org.freedesktop.portal.InputCapture.xml:
```diff
@@ -86,6 +86,11 @@
           Bitmask of requested capabilities, see the SupportedCapabilities property.
           This value is required and must not be zero.
 
+        * draft (b)
```
```suggestion
        * ``draft`` (b)
```
I think, just check with the others please.

sithlord48 commented:
If i could get some further guidance on this i added the draft but i am not sure what other changes in the code / xml files are needed.

jadahl commented:
> IOW it's just a boolean shoved into the options a{sv}.

What I found awkward is that it goes the whole "Request/Response" dance with nonsensical arguments when it could just be dumb method call, followed by the RequestResponse dance. It's also not unthinkable we might want to distinguish between options used when creating the session and initializing, which would be a bit messy if they were mixed between a Create and Init method. The alternative, deprecating CreateSession I'd imagine is something like:

```xml
    <!--
        CreateDraftSession:
        @options: Vardict with optional further information
        @session_handle: Object path for the created :ref:`org.freedesktop.portal.Session` object

        Creates a draft capture input session. A successfully created session can at
        any time be closed using :ref:`org.freedesktop.portal.Session.Close`, or may
        at any time be closed by the portal implementation, which will be
        signalled via :ref:`org.freedesktop.portal.Session::Closed`.

        Supported keys in the @options vardict include:

        * ``session_handle_token`` (``s``)

          A string that will be used as the last element of the session handle. Must be a valid
          object path element. See the :ref:`org.freedesktop.portal.Session` documentation for
          more information about the session handle.

        This method was added in version 2 of this interface.
    -->
    <method name="CreateDraftSession">
      <annotation name="org.qtproject.QtDBus.QtTypeName.In1" value="QVariantMap"/>
      <arg type="a{sv}" name="options" direction="in"/>
      <arg type="o" name="session_handle" direction="out"/>
    </method>
    <!--
        InitSession:
        @parent_window: Identifier for the application window, see :doc:`window-identifiers`
        @options: Vardict with optional further information
        @session_handle: Object path for the :ref:`org.freedesktop.portal.Request` object representing this call

        Initializes a capture input session. This may result the portal backend querying the user whether
        they want to allow the input of the session to be captured or not, and what capabilities to support.

        Supported keys in the @options vardict include:

        * ``handle_token`` (``s``)

          A string that will be used as the last element of the @handle. Must be a valid
          object path element. See the :ref:`org.freedesktop.portal.Request` documentation for
          more information about the @handle.

        * ``capabilities`` (``u``)

          Bitmask of requested capabilities, see the SupportedCapabilities property.
          This value is required and must not be zero.

        The following results get returned via the :ref:`org.freedesktop.portal.Request::Response` signal:

        * ``capabilities`` (``u``)

          The capabilities available to this session. This is always a
          subset of the requested capabilities.
          See the SupportedCapabilities property for details. Note that
          while a capability may be available to a session, there is no
          guarantee a device with that capability is currently available
          or if one does become available that it will trigger input capture.

          It is best to view this set as a negative confirmation - a
          capability that was requested but is missing is an indication that
          this application may not capture events of that capability.

        This method was added in version 2 of this interface.
    -->
    <method name="InitSession">
      <arg type="s" name="parent_window" direction="in"/>
      <annotation name="org.qtproject.QtDBus.QtTypeName.In1" value="QVariantMap"/>
      <arg type="a{sv}" name="options" direction="in"/>
      <arg type="o" name="handle" direction="out"/>
    </method>
    <!--
        CreateSession:
        @parent_window: Identifier for the application window, see :doc:`window-identifiers`
        @options: Vardict with optional further information
        @handle: Object path for the :ref:`org.freedesktop.portal.Request` object representing this call

        A short hand for calling :ref:`org.freedesktop.portal.InputCapture.CreateDraftSession()`
        followed by :ref:`org.freedesktop.portal.InputCapture.InitSession()`. See
        the corresponding methods for allowed options, and
        :ref:`org.freedesktop.portal.InputCapture.InitSession()` for the contents of the
        :ref:`org.freedesktop.portal.Request::Response` signal.

        Since version 2, this method is deprecated.
    -->
    <method name="CreateSession">
      <annotation name="org.freedesktop.DBus.Deprecated" value="true"/>
      <arg type="s" name="parent_window" direction="in"/>
      <annotation name="org.qtproject.QtDBus.QtTypeName.In1" value="QVariantMap"/>
      <arg type="a{sv}" name="options" direction="in"/>
      <arg type="o" name="handle" direction="out"/>
    </method>
```

whot commented:
> The alternative, deprecating CreateSession I'd imagine is something like

without my brain fully engaged yet: this LGTM :)


> I have no idea what exactly i need todo here for that.

copy the new methods CreateDraftSession and Init to the XML and then implement the hooks/calls to it if a client registers version 2 of the protocol. Basically grep for anything CreateSession does and copy it for the draft session.

jadahl commented:
>  @whot just to clarify i need to make new methods in the input_capture.h named init and create_draft_session

More or less, I'd try to restructure things so that what `create_session` currently does (except the D-bus method call / Request plumbing) is split up in two helpers, one that creates the session, one that initializes. Then `create_session0` call both, `create_draft_session()` only calls the first and `init()` calls the second.

sithlord48 commented:
Quick question for the xml changes do i add them to both just the impl or the non impl ?

whot commented:
> Quick question for the xml changes do i add them to both just the impl or the non impl ?

Both non-impl and impl

sithlord48 commented:
updated the xml files to what i think is correct.

KoljaFrahm commented:
So is something still missing? I hope this can be merged soon!

sithlord48 commented:
> So is something still missing? I hope this can be merged soon!

Yeah the code is missing then portal implementations have to also use v2 of the spec. I'm just focused on getting the xml first since those names are used in the code and that is the part im least sure of.

whot commented on data/org.freedesktop.portal.InputCapture.xml:
```diff
@@ -56,44 +56,55 @@
+        CreateDraftSession:
         @options: Vardict with optional further information
-        @handle: Object path for the :ref:`org.freedesktop.portal.Request` object representing this call
+        @session_handle: Object path for the created :ref:`org.freedesktop.portal.Session` object
```
This may be intentional or it may be a misunderstanding how the requests work, I'm not sure :) Are you intentionally making this *not* go through a Request here? 

Makes sense since there's no reason this can fail or be delayed so requests aren't necessary but let's make sure it's spelled out in the commit message at least.

@jadahl - do you prefer this via Request for consistency with the others?

sithlord48 replied:
I literally have no idea what i am doing. I was told on the matrix that support was as easy as copying over the methods used in remote desktop. But after i was told here in review to add these new methods etc.. So im just trying todo what was asked to get this working so we can use clipboard with deskflow, input-leap etc... 

If you or anyone who knows what they are doing wants to take this over im ok with that also . Otherwise ill keep chugging along until it gets there (just maybe alot more questions along the way)

jadahl replied:
I don't think we need to go through Request when creating a draft session.

whot commented on data/org.freedesktop.portal.InputCapture.xml:
```diff
@@ -56,44 +56,55 @@
+        InitSession:
+        @parent_window: Identifier for the application window, see :doc:`window-identifiers`
+        @options: Vardict with optional further information
+        @session_handle: Object path for the :ref:`org.freedesktop.portal.Request` object representing this call
```
```suggestion
        @handle: Object path for the :ref:`org.freedesktop.portal.Request` object representing this call
```
This one needs to go through a request.

The TLDR regarding requests: dbus times out faster than the user clicks so the workaround for the portals is that calls immediately return the object path for a temporary object (a `Request`) and that request then triggers a signal when the user is done. That request signal is the `Response` and contains the data we actually one (the `the following results get returned...` bit). The doc has a bit more details but the short summary is: anything that requires user interaction needs to go through a Request.

This doesn't apply to the impls since we can control the timeouts there so they don't need the request dance.

sithlord48 replied:
Updated

whot commented on data/org.freedesktop.portal.InputCapture.xml:
```diff
@@ -106,14 +117,46 @@
+
+          The capabilities available to this session. This is always a
+          subset of the requested capabilities.
+    -->
```
Removing the `<method name="CreateSession">` is an API break, we need to keep that one around as-is.

sithlord48 replied:
I replaced the method in both XML files.

whot commented on src/input-capture.c:
```diff
@@ -52,7 +52,7 @@ static InputCapture *input_capture;
 static GQuark quark_request_session;
 
-GType input_capture_get_type (void);
+GType input_capture_get_type (void) G_GNUC_CONST;
```
Unrelated, should be a separate commit please, same with the clipboard hunks here which are a separate thing too.

sithlord48 replied:
Just to make sure i have this correct  these are the two commits you want ? 
1. commit that adds V2 methods 
2. commit that adds the clipboard chunks

jadahl replied:
I think the point was to make the change about `G_GNUC_CONST` its own commit, since it's unrelated to everything else.

whot commented on src/input-capture.c:
```diff
@@ -31,7 +31,7 @@
 #include "xdp-utils.h"
 
-#define VERSION_1  1 /* Makes grep easier */
+#define VERSION  2 /* Makes grep easier */
```
```suggestion
#define VERSION_1 1 /* Makes grep easier */
#define VERSION_2 2
```
And then use `VERSION_2` where we want that newer version.

jadahl commented on data/org.freedesktop.impl.portal.InputCapture.xml:
```diff
@@ -31,6 +31,119 @@
+          The capabilities available to this session. This is always a
+          subset of the requested capabilities.
+    -->
+    <method name="CreateSession">
```
There is already `CreateSession` in this interface, so no need to add this.

sithlord48 replied:
oh i see thanks updating soon.

jadahl commented on data/org.freedesktop.portal.InputCapture.xml:
```diff
@@ -113,7 +113,107 @@
+    </method>
+
+    <!--
+        CreateSession:
```
There is already documentation for `CreateSession` at line 59.

sithlord48 replied:
noticed and removed the duplicate

sithlord48 commented:
@jadahl  @whot is the XML correct now ?

whot commented on data/org.freedesktop.portal.InputCapture.xml:
```diff
@@ -114,6 +114,78 @@
+        @options: Vardict with optional further information
+        @handle: Object path for the :ref:`org.freedesktop.portal.Request` object representing this call
+
+        Initializes a capture input session. This may result the portal backend querying the user whether
```
```suggestion
        Initializes a capture input session previously created with CreateDraftSession. This may result the portal backend querying the user whether
```

sithlord48 replied:
Changed the description to 
```
 This method may only be called once and only on a session previously created with CreateDraftSession.
 Initializes a capture input session previously created with CreateDraftSession. This may cause
 the portal backend to query the user whether they want to allow the input of the session to be
 captured or not, and what capabilities to support.
```

whot commented on data/org.freedesktop.portal.InputCapture.xml:
```diff
@@ -114,6 +114,78 @@
+        @handle: Object path for the :ref:`org.freedesktop.portal.Request` object representing this call
+
+        Initializes a capture input session. This may result the portal backend querying the user whether
+        they want to allow the input of the session to be captured or not, and what capabilities to support.
```
For the more final versions we should also add things like:
```
This method may only be called once and only on a session previously 
created with CreateDraftSession.
```
and possibly

```
Clients implementing version 2 of this interface should prefer CreateDraftSession 
followed by InitSession over the version 1 CreateSession method.
```

sithlord48 replied:
I added the second note at the bottom of the createSession area line 70

whot reviewed (commented):
The XML looks good now, ready for implementing :)

sithlord48 commented:
> The XML looks good now, ready for implementing :)

I changed the handle_create_session to be handle_init_session and added a for now empty
`handle_create_draft_session` as im not sure really what the return or input in there i also modified the old `handle_create_session` to call handle draft then init session

Thanks for your help and input I think we are getting close now.

sithlord48 commented:
Oh i broke a test but not sure why or what it was testing either :(

sithlord48 commented:
Every thing is updated, Tests no pass locally , gonna spin the review buttons

swick commented on data/org.freedesktop.portal.InputCapture.xml:
```diff
@@ -114,6 +117,80 @@
+        This method was added in version 2 of this interface.
+    -->
+    <method name="CreateDraftSession">
+      <annotation name="org.qtproject.QtDBus.QtTypeName.In1" value="QVariantMap"/>
```
```suggestion
      <annotation name="org.qtproject.QtDBus.QtTypeName.In0" value="QVariantMap"/>
```

sithlord48 commented:
> Only test the version clipboard integration is added.

ok

> Perhaps do more or less what the remote desktop clipboard tests does?

You know i should have realized this

sithlord48 commented on src/input-capture.c:
```diff
@@ -226,8 +254,18 @@ static XdpOptionKey input_capture_create_session_options[] = {
+                                         const char            *arg_parent_window,
+                                         GVariant              *arg_options)
+{
+  return FALSE;
```
Can anyone tell me what part of the old handle_session happens here now ?

jadahl commented on src/input-capture.c:
```diff
@@ -226,8 +254,18 @@ static XdpOptionKey input_capture_create_session_options[] = {
+
 static gboolean
-handle_create_session (XdpDbusInputCapture   *object,
+handle_init_session (XdpDbusInputCapture   *object,
```
The function below should call `xdp_dbus_impl_input_capture_call_create_session()` and filter the options using a new `input_capture_init_session_options` array.

sithlord48 replied:
Where are these defined ?

sithlord48 commented on data/org.freedesktop.portal.InputCapture.xml:
```diff
@@ -114,6 +117,80 @@
+    </method>
+
+    <!--
+        InitSession:
```
Is this not the same as this (in the options) 
```
        * ``handle_token`` (``s``)

          A string that will be used as the last element of the @handle. Must be a valid
          object path element. See the :ref:`org.freedesktop.portal.Request` documentation for
          more information about the @handle.
```

sithlord48 commented:
I think i got those all @jadahl can you check the xml is correct if so ill work to update the code

jadahl commented on data/org.freedesktop.portal.InputCapture.xml:
```diff
@@ -114,6 +117,88 @@
+    -->
+    <method name="InitSession">
+      <arg type="s" name="parent_window" direction="in"/>
+      <annotation name="org.qtproject.QtDBus.QtTypeName.In1" value="QVariantMap"/>
```
With the session handled argument added, this `In1` will be `In2`.

jadahl reviewed (commented):
Getting there!

jadahl commented on data/org.freedesktop.impl.portal.InputCapture.xml:
```diff
@@ -72,6 +72,94 @@
+    -->
+    <method name="InitSession">
+      <arg type="o" name="session_handle" direction="in"/>
+      <arg type="o" name="handle" direction="in"/>
```
the order is different here from the other session methods (e.g. GetZones), here it's handle, session_handle, but below it's session_handle, handle. It seems the other methods below it's "handle, session_handle".

jadahl reviewed (commented):
Just a tiny thing left, (argument order), but other than that, the XML looks good now!

sithlord48 commented:
Updated w/ the args @jadahl  They should be correct in code now (at least the args) . 
 I am still unsure how the code gets split into the two new methods
